### PR TITLE
JAX transformer text classification

### PR DESCRIPTION
**Pull request by mtsokol**
This PR adds a tutorial with JAX transformer text classification task on IMDB dataset. It follows one of the [Keras tutorials](https://keras.io/examples/nlp/text_classification_with_transformer/). It's still missing the narrative.

@vfdev-5 I took most of the training, validation and visualization code from https://github.com/jax-ml/jax-ai-stack/pull/95.

I perform dataset preprocessing in a separate function in order to show usage of `grain.RandomAccessDataSource`.

### CODE REVIEW

File: docs/JAX_transformer_text_classification.md [hunk 1]
```diff
@@ -0,0 +1,355 @@
...
+import tqdm
+import matplotlib.pyplot as plt
+import numpy as np
+import requests
+```
+
+## Dataset
+
+```{code-cell} ipython3
+def prepare_imdb_dataset(num_words: int) -> tuple:
```
Comment (mtsokol) [inline docs/JAX_transformer_text_classification.md:35]: A few bits here are taken from https://github.com/keras-team/keras/blob/495a570c5350e689d686c572009811cc4492d8c6/keras/src/datasets/imdb.py#L13 to fully reproduce data setup of [a Keras tutoral](https://keras.io/examples/nlp/text_classification_with_transformer/). I will simplify it.

Comment (mtsokol) [reply docs/JAX_transformer_text_classification.md]: I revisited dataset preprocessing. Done!

File: docs/JAX_transformer_text_classification.md [hunk 2]
```diff
@@ -0,0 +1,375 @@
...
+---
+
+# Text classification with transformer model
+
+In this short tutorial, we will perform sentiment analysis on movie reviews. We would like
+to know if a review is overall positive or negative, so we're facing a classification task.
+
+For that purpose, we're going to build a model comprised of a transformer and dense layers.
+What is special about transformers is that they utilize attention mechanism that allows
+the model to focus on the most relevant parts of the input sequence when processing information.
```
Comment (melissawm) [inline docs/JAX_transformer_text_classification.md:22]:
```suggestion
What is special about transformers is that they utilize an attention mechanism that allows
the model to focus on the most relevant parts of the input sequence when processing information.
```

File: docs/JAX_transformer_text_classification.md [hunk 3]
```diff
@@ -0,0 +1,375 @@
...
+vocab_size = 20000  # Only consider the top 20k words
+maxlen = 200  # Only consider the first 200 words of each movie review
+(x_train, y_train), (x_test, y_test) = prepare_imdb_dataset(num_words=vocab_size)
+print(len(x_train), "Training sequences")
+print(len(x_test), "Validation sequences")
+x_train = pad_sequences(x_train, max_len=maxlen)
+x_test = pad_sequences(x_test, max_len=maxlen)
+```
+
+As our dataset are relatively small NumPy arrays we define it as a `RandomAccessDataSource`:
```
Comment (melissawm) [inline docs/JAX_transformer_text_classification.md:115]: Maybe link to the PyGrain docs?
```suggestion
As our dataset are relatively small NumPy arrays we define it as a `RandomAccessDataSource` (see [PyGrain Data Sources](https://github.com/google/grain/blob/main/docs/data_sources.md) for more details):
```

File: docs/JAX_transformer_text_classification.md [hunk 4]
```diff
@@ -0,0 +1,375 @@
...
+        return x
+```
+
+```{code-cell} ipython3
+model = MyModel(rngs=nnx.Rngs(0))
+nnx.display(model)
+```
+
+## Training
+
```
Comment (melissawm) [inline docs/JAX_transformer_text_classification.md:254]: Can we add a few words about using nnx.Optimizer and the other nnx objects for training here? I think it's worth highlighting

File: docs/JAX_transformer_text_classification.md [hunk 5]
```diff
@@ -0,0 +1,375 @@
...
+```{code-cell} ipython3
+fig, axs = plt.subplots(1, 2, figsize=(10, 5))
+axs[0].set_title("Loss value on test set")
+axs[0].plot(eval_metrics_history["test_loss"])
+axs[1].set_title("Accuracy on test set")
+axs[1].plot(eval_metrics_history["test_accuracy"])
+```
+
+As we can se we acquired over 85% accuracy after 10 epochs. From the loss plot
+we see that in the last few ones loss value didn't improve much.
```
Comment (melissawm) [inline docs/JAX_transformer_text_classification.md:375]:
```suggestion
As we can see, we acquired over 85% accuracy after 10 epochs. From the loss plot,
we see that in the last few epochs the loss value didn't improve much.
```

Comment (melissawm) [reply docs/JAX_transformer_text_classification.md:375]: Does this mean we should have chosen less epochs?

**Review by melissawm — COMMENTED** [review]
Couple of notes:
- It may be worth adding a few words about the Grain usage.
- I would add some "Next steps" notes, either finalizing with an interpretation of the results (why does our model work? Could we have chosen another one? What was the advantage in using the jax-ai-stack here?)

I have another more holistic comment for @jakevdp : do we want to be more explicit in these tutorials about which tools we are using are part of the jax-ai-stack? As an external reader, it is not obvious to me and I think we could do a better job of highlighting why the jax ecosystem tools are helpful in the examples.

File: docs/JAX_transformer_text_classification.md (hunk 4 above)
Comment (mtsokol) [reply docs/JAX_transformer_text_classification.md:272]: Done! I added two sentences about training.

File: docs/JAX_transformer_text_classification.md (hunk 3 above)
Comment (mtsokol) [reply docs/JAX_transformer_text_classification.md]: Done!

File: docs/JAX_transformer_text_classification.md (hunk 2 above)
Comment (mtsokol) [reply docs/JAX_transformer_text_classification.md]: Done!

File: docs/JAX_transformer_text_classification.md (hunk 5 above)
Comment (mtsokol) [reply docs/JAX_transformer_text_classification.md]: IMO it means we converged - I expanded this section a bit.

**Comment by mtsokol** [comment]
> It may be worth adding a few words about the Grain usage.

Sure! I added a few sentences about Grain.

> I would add some "Next steps" notes, either finalizing with an interpretation of the results (why does our model work? Could we have chosen another one? What was the advantage in using the jax-ai-stack here?)

I expanded the last section a bit with what came to my mind - let me know if it should be expanded further.

> do we want to be more explicit in these tutorials about which tools we are using are part of the jax-ai-stack?

I added `Tools overview` section with a few links to the most relevant libraries.

**Review by melissawm — APPROVED** [review]
Looks great, thanks @mtsokol !

**Comment by trallard** [comment]
@jakevdp this should be ready now 🚀

File: docs/JAX_transformer_text_classification.md [hunk 6]
```diff
@@ -0,0 +1,402 @@
...
+
+```{code-cell} ipython3
+# Required packages
+# !pip install -U jax flax optax
+# !pip install -U grain tqdm requests matplotlib
+```
+
+## Tools overview
+
+Here's a list of key packages that belong to JAX AI stack:
```
[the remark this answers is not in the document]
Comment (mtsokol) [reply docs/JAX_transformer_text_classification.md]: Done!

File: docs/JAX_transformer_text_classification.md [hunk 7]
```diff
@@ -0,0 +1,402 @@
...
+plt.legend()
+```
+
+```{code-cell} ipython3
+fig, axs = plt.subplots(1, 2, figsize=(10, 5))
+axs[0].set_title("Loss value on test set")
+axs[0].plot(eval_metrics_history["test_loss"])
+axs[1].set_title("Accuracy on test set")
+axs[1].plot(eval_metrics_history["test_accuracy"])
+```
```
Comment (jakevdp) [inline docs/JAX_transformer_text_classification.md:395]: Could we somehow print a selection of (truncated) test samples and labels? Without that it's hard for the reader to tell what exactly the model is doing.

**Review by jakevdp — CHANGES_REQUESTED** [review]

File: docs/JAX_transformer_text_classification.md [hunk 8]
```diff
@@ -0,0 +1,402 @@
...
+        x = self.dense1(x)
+        x = jax.nn.relu(x)
+        x = self.dropout2(x)
+        x = jax.nn.softmax(x)
+        return x
+```
+
+```{code-cell} ipython3
+model = MyModel(rngs=nnx.Rngs(0))
+nnx.display(model)
```
[the remark this answers is not in the document]
Comment (mtsokol) [reply docs/JAX_transformer_text_classification.md:270]: Done!

File: docs/JAX_transformer_text_classification.md (hunk 7 above)
Comment (mtsokol) [reply docs/JAX_transformer_text_classification.md:397]: Done! I added a cell at the end where I inspect a few decoded samples together with a predicted and an actual label.

I squashed the PR, it's ready from my side!

**Review by jakevdp — APPROVED** [review]
Thanks!
